std/digest/hmac.d

```diff
@@ -0,0 +1,323 @@
+        import std.string : representation;
+        ubyte[] data1, data2;
+        auto hmac = HMAC!SHA1("My s3cR3T keY".representation);
+        hmac.put(data1);
+        hmac.start();    // reset digest
+        hmac.put(data2); // start over
+    }
+
+    /**
+     * Feeds a piece of data into the hash computation. This method allows the
+     * type to be used as an $(XREF range, OutputRange).
+     *
+     * Returns:
+     * A reference to the digest for convenient chaining.
+     */
+
+    ref put(in ubyte[] data...) return
+    {
+        digest.put(data);
+        return this;
+    }
+
+    ///
+    @safe pure nothrow @nogc unittest
+    {
+        import std.digest.sha, std.digest.hmac;
+        import std.string : representation;
+        ubyte[] data1, data2;
+        auto hmac = HMAC!SHA1("My s3cR3T keY".representation);
+        hmac.put(data1)
+            .put(data2);
+    }
+
+    /**
+     * Resets the digest and returns the finished hash.
+     */
+
+    DigestType!H finish()
+    {
+        ubyte[blockSize / 8] opad = void;
+        foreach(immutable i; 0 .. blockSize / 8)
+            opad[i] = key[i] ^ 0x5c;
+
+        auto tmp = digest.finish();
+
+        digest.start();
+        digest.put(opad[]);
+        digest.put(tmp);
+        auto result = digest.finish();
+        start();    // reset the digest
+        return result;
+    }
+
+    ///
+    @safe pure nothrow @nogc unittest
+    {
+        import std.digest.sha, std.digest.hmac;
+        import std.string : representation;
+        ubyte[] data1, data2;
+        auto hmac = HMAC!SHA1("My s3cR3T keY".representation);
+        auto digest = hmac.put(data1)
+                          .put(data2)
+                          .finish();
+    }
+}
+
+template hmac(H)
+if(isDigest!H && hasBlockSize!H)
+{
+    alias hmac = hmac!(H, H.blockSize);
+}
+
+template hmac(H, size_t blockSize)
+if(isDigest!H)
+{
+    /**
+     * Convenience function for constructing an HMAC digest with the
+     * specified secret.
+     *
+     * Returns:
+     * An instance of HMAC that can be fed data as desired, and finished
+     * to compute the final hash when done.
+     */
+
+    auto hmac(scope const(ubyte)[] secret)
+    {
+        return HMAC!(H, blockSize)(secret);
+    }
+
+    ///
+    @safe pure nothrow @nogc unittest
+    {
+        import std.digest.sha, std.digest.hmac;
+        import std.string : representation;
+        ubyte[] data1, data2;
+        auto digest = hmac!SHA1("My s3cR3T keY".representation)
+                          .put(data1)
+                          .put(data2)
+                          .finish();
+    }
+
+    /**
+     * Convenience function for computing an _HMAC digest over the given
+     * ranges of data with the specified secret.
+     *
+     * Returns:
+     * The final _HMAC hash.
+     */
+
+    DigestType!H hmac(Range)(auto ref Range range, scope const(ubyte)[] secret)
+    if(!isArray!Range && isDigestibleRange!Range)
```

Comment:
Don't know. This is an adaptation of `std.digest.digest.digest()`: https://github.com/D-Programming-Language/phobos/blob/master/std/digest/digest.d#L405 and https://github.com/D-Programming-Language/phobos/blob/master/std/digest/digest.d#L430

I'd have to test whether they can be merged into one function.